python/lsst/ci/build.py

```diff
@@ -345,6 +393,144 @@ def run(args):
 
         b = Builder(build_dir, manifest, progress, eups_obj)
         b.rm_status()
-        retcode = b.build()
-        b.write_status()
-        sys.exit(retcode == 0)
+
+        try:
+            if agent == "unknown_agent":
+                raise UnknownAgentError(agent)
+
+            # Verify PR info and post pending Github status
+            if pr_info.items:
+                post_statuses.post_all_status(
+                    pr_info, agent=agent, state="pending", description=f"Build started on {agent}"
+                )
+
+            # Attempt the build
+            retcode = b.build()
+
+        except UnknownAgentError as agent_err:
+            print(f"WARNING: {agent_err}")
+            retcode = b.build()
+
+        except Exception as other_ex:
+            print(f"Build failed on {agent}: {other_ex}")
+            traceback.print_exc()
+            retcode = False
+
+        finally:
+            # Ensure status is always written
+            b.write_status()
+
+            # Final status posting based on retcode
+            if retcode:
+                if pr_info.items:
+                    post_statuses.post_all_status(
+                        pr_info, agent=agent, state="success", description=f"Build succeeded on {agent}"
+                    )
+                else:
+                    print(f"Build succeeded on {agent}.")
+
+                sys.exit(0)
+            else:
+                if pr_info.items:
+                    post_statuses.post_all_status(
+                        pr_info, agent=agent, state="failure", description=f"Build failed on {agent}"
+                    )
+                else:
+                    print(f"Build failed on {agent}")
+                sys.exit(1)
+
+
+class UnknownAgentError(Exception):
+    """
+    Raised when the agent is unknown.
```

Comment:
General comment that the summary should be on the same line as the opening quotes.